### PR TITLE
Add logging support to Cast Propagation transformation from python

### PR DESCRIPTION
**Pull request by satyajandhyala**
**Description**: Added logging and other miscellaneous changes

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: orttraining/orttraining/core/framework/ortmodule_graph_builder.h [hunk 1]
```diff
@@ -28,7 +28,11 @@ struct OrtModuleGraphBuilderConfiguration {
   bool use_invertible_layernorm_grad = false;
   bool build_gradient_graph = true;
 
+  // Graph transformer configuration
   TrainingSession::TrainingConfiguration::GraphTransformerConfiguration graph_transformer_config{};
+
+  // Log severity
+  logging::Severity loglevel{logging::Severity::kWARNING};
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/framework/ortmodule_graph_builder.h:35]: see session_option.h, we can use int type here, and skip the trouble for pybind. 

`  int session_log_severity_level = -1;`

File: orttraining/orttraining/python/training/_ortmodule_graph_execution_manager.py [hunk 2]
```diff
@@ -272,5 +272,10 @@ def _initialize_graph_builder(self, training):
         grad_builder_config.graph_transformer_config = C.GraphTransformerConfiguration()
         grad_builder_config.graph_transformer_config.propagate_cast_ops_level = self._propagate_cast_ops_level
         grad_builder_config.graph_transformer_config.propagate_cast_ops_allow = self._propagate_cast_ops_allow
+        grad_builder_config.loglevel = {_logger.LogLevel.VERBOSE : C.Severity.VERBOSE,
```
Comment (SherlockNoMad) [inline orttraining/orttraining/python/training/_ortmodule_graph_execution_manager.py:275]: Can we just pass the self._loglevel to the backend? After all, it just an integer.  
We can skip the pybind for "Severity" and make the frontend code simple and clean.

Comment (satyajandhyala) [reply orttraining/orttraining/python/training/_ortmodule_graph_execution_manager.py:275]: I think this is more type safe and Thiago also prefers using typesafe data than int, I think.

File: orttraining/orttraining/core/framework/ortmodule_graph_builder.h (hunk 1 above)
Comment (satyajandhyala) [reply orttraining/orttraining/core/framework/ortmodule_graph_builder.h:35]: please see the reply below.

File: orttraining/orttraining/python/training/_ortmodule_graph_execution_manager.py (hunk 2 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/_ortmodule_graph_execution_manager.py:275]: If we can make the onnxruntime Python API more idiomatic, it is preferable. Friendlier APIS make it more likely to be adopted by Python community users

Using ints to replace enums is kind of low level C API and might feel weird for Python users, especially the ones that never worked with C/C++ before

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 3]
```diff
@@ -520,14 +520,23 @@ py::class_<TrainingAgent>(m, "TrainingAgent", R"pbdoc(This is the main class use
   py::class_<OrtModuleGraphBuilderConfiguration> module_graph_builder_config(
       m, "OrtModuleGraphBuilderConfiguration",
       R"pbdoc(Configuration information for module graph builder.)pbdoc");
+
+  py::enum_<Severity>(m, "Severity", py::arithmetic(), py::module_local())
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:524]: one improvement point is to check if there are existing APIs in this file which could leverage this new enum

Comment (satyajandhyala) [reply orttraining/orttraining/python/orttraining_pybind_state.cc:524]: In this file I don't see where I can use the new enu.

Comment (satyajandhyala) [reply orttraining/orttraining/python/orttraining_pybind_state.cc:524]: OrtRunOptions also is something that can change data from from int to OrtLoggingLevel.

**Comment by SherlockNoMad** [comment]
let's start adding labels for future PRs.

File: onnxruntime/core/optimizer/propagate_cast_ops.cc [hunk 4]
```diff
@@ -130,12 +130,22 @@ static Status RemoveCastNodesChain(Graph& graph, std::vector<Node*> casts, std::
...
   if (producer) {
-    int input_index = optimizer_utils::IndexOfNodeInput(*lead_cast, *cast_input);
-    graph.RemoveEdge(producer->Index(), lead_cast->Index(), output_index, input_index);
-    if (consumers.empty()) {
-      auto& outputs = producer->MutableOutputDefs();
-      std::replace(outputs.begin(), outputs.end(), cast_input, cast_output);
-      graph.UpdateProducerNode(cast_output->Name(), producer->Index());
+    if (graph.IsOutput(cast_output)) {
+      // cast_output is a graph output. Replace the cast node with an Identity operator unless node
+      // has other outputs.
```
Comment (satyajandhyala) [inline onnxruntime/core/optimizer/propagate_cast_ops.cc:141]: This comment needs to be fixed as " ... unless  node has no other outputs."

**Review by SherlockNoMad — APPROVED** [review]

**Review by thiagocrepaldi — APPROVED** [review]

**Pull request merged**